bsr_tensor_invariants.md

```diff
@@ -0,0 +1,258 @@
+| Author     | Pearu Peterson  |
+| Created    | 2022-05-06      |
+
+The aim of this blog post is to define the invariants of PyTorch
+tensors with BSR layout. As a first instance, we assume that a BSR
+tensor has zero batch dimensions and is a non-hybrid tensor (tensor
+values are scalars).
+
+## Sparse tensor BSR format
+
+A sparse tensor BSR format is a generalization of sparse tensor CSR
+format where BSR indices refer to 2-dimensional dense blocks in the
+tensor rather than elements as in the case of a CSR tensor. For
+example:
+```python
+>>> dense = torch.tensor([[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 17], [18, 19, 20, 21, 22, 23]])
+>>> bsr = dense.to_sparse_bsr(blocksize=(2, 3))
+>>> bsr
+tensor(crow_indices=tensor([0, 2, 4]),
+       col_indices=tensor([0, 1, 0, 1]),
+       values=tensor([[[ 0,  1,  2],
+                       [ 6,  7,  8]],
+
+                      [[ 3,  4,  5],
+                       [ 9, 10, 11]],
+
+                      [[12, 13, 14],
+                       [18, 19, 20]],
+
+                      [[15, 16, 17],
+                       [21, 22, 23]]]), size=(4, 6), nnz=4,
+       layout=torch.sparse_bsr)
+```
+Notice that `crow_indices` and `col_indices` contain indices of a CSR
+tensor with shape `(2, 2)` while the shape of BSR tensor is `(4, 6)`
+because it incorporates the shape of `values` blocks: a block is a
+strided tensor with shape `(2, 3)`.
+
+### BSR tensor members
+
+A tensor with BSR layout has the following members (as defined by constructor `sparse_bsr_tensor`):
+- `crow_indices` contains the compressed block row indices information
+- `col_indices` contains block column indices
+- `values` contains the values of tensor elements organized in a 3-dimensional array of blocks
+- `size` defines the shape of tensor
+- `dtype` defines the dtype of tensor elements
+- `layout` holds the layout parameter
+- `device` holds the device of values storage
+- `pin_memory` defines if cuda storage uses pinned memory
+
+### Type invariants
+
+1.1 `crow_indices.dtype == indices_dtype`
+
+1.2 `col_indices.dtype = indices_dtype`
+
+1.3 `indices_dtype` is `int32` (default) or `int64`
+
+1.4 `values.dtype == dtype`
+
+1.5 `dtype` is `float32` (default), `float64`, `int8`, `int16`,
+    `int32`, `int64`, `bool`, `complex32` (unsupported ATM),
+    `complex64`, or `complex128`.
+
+### Layout invariants
+
+2.1 `crow_indices.layout == torch.strided`
+
+2.2 `col_indices.layout == torch.strided`
+
+2.3 `values.layout == torch.strided`
+
+2.4 `layout == torch.sparse_bsr`
+
+### Shape and strides invariants
+
+3.1 `size == (nrows * blocksize[0], ncols * blocksize[1])`, that is, a
+    BSR tensor represents a two dimensional tensor.
+
+3.2 `crow_indices.dim() == 1`
+
+3.3 `col_indices.dim() == 1`
+
+3.4 `values.dim() == 3`
+
+3.5 `crow_indices.is_contiguous()`
+
+3.6 `col_indices.is_contiguous()`
+
+3.7 `values.is_contiguous()` or `values.transpose(-2, -1).is_contiguous()`
```

Comment:
I agree that this is an important condition if we want to support transposition as a view.